utils/form_validation.php

```diff
@@ -0,0 +1,310 @@
+     * @param string $attrs
+     * @return string
+     */
+    function render(string $attrs = "")
+    {
+        return "<label for='" . $this->key . "'>"
+            . $this->label
+            . "<input ${attrs} name=\"$this->key\" id=\"$this->key\" value=\"$this->value\" " . ($this->valid() ? "" : " aria-invalid=true") . "></label>";
+    }
+
+    /** Adds a validation error */
+    protected function add_error($err)
+    {
+        $this->errors[] = $err;
+    }
+
+    function valid()
+    {
+        return empty($this->errors);
+    }
+
+    /** Outputs a variable (Decorator pattern) */
+    function out(&$var)
+    {
+        $var = $this->value;
+        return $this;
+    }
+
+    /** Helper method to skip logic when $_POST is empty */
+    protected function skip()
+    {
+        return $this->context->empty;
+    }
+
+    /** Makes the field required */
+    function required(string $msg = null)
+    {
+        if ($this->skip())
+            return $this;
+        if (!$this->value) {
+            $this->add_error($msg ?? "Requis");
+        }
+        return $this;
+    }
+
+
+    /** Enables Decorator pattern */
+    function number(string $key, string $msg = "")
+    {
+        return $this->context?->number($key, $msg) ?? new NumberField($key);
+    }
+    /** Enables Decorator pattern */
+    function string(string $key, string $msg = "")
+    {
+        return $this->context?->string($key, $msg) ?? new StringField($key);
+    }
+    /** Enables Decorator pattern */
+    function date(string $key, string $msg = "")
+    {
+        return $this->context?->date($key, $msg) ?? new DateField($key);
+    }
+}
+
+class NumberField extends Field
+{
+    function check($msg = null)
+    {
+        if (!preg_match("/^[\d]*$/", $this->value)) {
+            $this->add_error($msg ?? "Format invalide");
+        }
+    }
+
+    function render(string $attrs = "")
+    {
+        return parent::render("type = number");
+    }
+
+    /** Set upper limit for the number field */
+    function max(int $count, string $msg = null)
+    {
+        if ($this->skip())
+            return $this;
+        if ($this->value > $count) {
+            $this->add_error($msg ?? "Trop grand");
+        }
+        return $this;
+    }
+
+    /** Set lower limit for the number field */
+    function min(int $count, string $msg = null)
+    {
+        if ($this->skip())
+            return $this;
+        if ($this->value < $count) {
+            $this->add_error($msg ?? "Trop petit");
+        }
+        return $this;
+    }
+}
+
+class DateField extends Field
+{
+    function check($msg = null)
+    {
+        if (!preg_match("/^[\d-]*$/", $this->value)) {
+            $this->add_error($msg ?? "Format invalide");
+        }
+    }
+
+    function render(string $attrs = "")
+    {
+        return parent::render("type = date");
+    }
+
+    /** Set upper date limit
+     * @todo better naming?
+     */
+    function before(string $date, string $msg = null)
```

Comment:
When I leave `$start_date` empty, there are two errors : "required" and "too soon".
I would prefer if there wasn't the "too soon" error. 
Maybe skip the checks (date, string length, int comparisons ...) if the field was not filled?

Reply:
I think I had a previous iteration without stacking checks. I will revert to that :)

Reply:
Done!